common/client-core/src/client/topology_control/smart_api_provider.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2024 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: GPL-3.0-only
+
+//! Caching, piecewise API Topology Provider
+//!
+
+#![warn(missing_docs)]
+
+use async_trait::async_trait;
+use log::{debug, error, warn};
+pub use nym_topology::providers::piecewise::Config;
+use nym_topology::{
+    providers::piecewise::{NymTopologyProvider, PiecewiseTopologyProvider},
+    EpochRewardedSet, NymTopology, RoutingNode, TopologyProvider,
+};
+use nym_validator_client::UserAgent;
+use rand::{prelude::SliceRandom, thread_rng};
+use url::Url;
+
+/// Topology Provider build around a cached piecewise provider that uses the Nym API to
+/// fetch changes and node details.
+#[derive(Clone)]
+pub struct NymApiTopologyProvider {
```

Comment:
What about using something like `NymApiSmartTopologyProvider`? Having two types with the same name was a bit confusing initially when reading the the code